Notify users that the page has been updated and incite a reload

mrmir opened the pull request:
https://community.openproject.org/projects/openproject/work_packages/56801

_____________________________

Documenting the new component and using it to replace streamable_flash_banner will be done in https://community.openproject.org/projects/openproject/work_packages/56802

oliverguenther committed:
Add new flash stream action

oliverguenther committed:
Use the stream action in a new MeetingsController#check_for_updates

oliverguenther committed:
Adapt action in frontend

oliverguenther committed:
Update the meeting when saving an agenda item

oliverguenther committed:
Use a timestamp to check if we should send a flash

oliverguenther committed:
Allow notification to be dismissed

mrmir committed:
Add i18n strings

mrmir committed:
Pacify eslint and rubocop

mrmir committed:
Refactor update flash component

mrmir committed:
Merge branch 'dev' into feature/meetings-update-flash

oliverguenther committed:
Remove the double touch of meeting

oliverguenther committed:
Update meeting instance variable to the one from the service

The @meeting variable was still showing the old timestamps, as only the association
got updated. This feels weird, but saves a #reload.

oliverguenther committed:
Write spec for flash notification to appear only in one window

mrmir commented on modules/meeting/app/components/meetings/update_flash_component.sass:
```diff
@@ -0,0 +1,4 @@
+@import 'helpers'
+
+.op-meeting-update-flash
```
Hi @HDinger, could I get your opinion on this? I want to override the default width to prevent a new line in the flash message. Is this an okay way to do that?

HDinger replied:
Hi @mrmir 

I see two potential issues with that:

1. I don't know the exact use case but when doing such an optimization to avoid line breaks, there is usally the question of other languages. Most likely there will be at least one language that is longer and thus the line will break nevertheless. Is that accepted? 
2. This is a very specific rule. However, I can imagine that there are other places that want a wider flash message. So maybe a generic `-wide` class would be better.

HDinger replied:
While thinking about it, I am wondering why there is the need to avoid line breaks overall. We have that already in other flashes, so why do we put extra effort here to get around that?

mrmir replied:
Hmm, I was just trying to get it to look like the specification, but the point about translations messing it up anyway is something I completely overlooked. I will leave it as default, thanks!

oliverguenther committed:
Calculate value to be compared instead of using updated_at

updated_at breaks when uploading attachments, as the same page will be
updated without us having a chance to update the header.

Instead, calculate an "etag" value from the meeting sections and agenda
items timestamps as well as the meeting's own lock_version.

mrmir committed:
Remove custom styling

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/meetings-update-flash

oliverguenther committed:
Update header to turbo stream

mrmir committed:
WIP specs

oliverguenther committed:
Allow tests to call poll trigger manually

This allows for a more stable check in tests, avoiding race conditions

mrmir committed:
Add feature flag

mrmir committed:
Add spec for meeting state

mrmir committed:
Reinstate the open_history_modal method

mrmir committed:
Reinstate retry block and pacify eslint

mrmir committed:
Merge branch 'dev' into feature/meetings-update-flash

mrmir committed:
Pacify rubocop

mrmir committed:
Fix reinstation of the open_history_modal method

mrmir commented:
@oliverguenther I'm still not sure if the specs are 100% indicative of the expected behaviour, so please take a close look at them again. 

 I also added `open_history_modal` back in as the alternative was not being used in multiple places anymore and so this seemed less clunky overall

toy commented on app/components/concerns/op_turbo/streamable.rb:
```diff
@@ -100,6 +100,10 @@ def component_wrapper(method = nil, tag: "div", **kwargs, &block)
 
+      def inline_actions
+        %i[dialog flash]
+      end
```
Is this meant to be overridable? Otherwise can be a constant

toy commented on modules/meeting/app/components/meetings/side_panel/state_component.html.erb:
```diff
@@ -26,7 +26,9 @@
-              data: { 'turbo-stream': true, 'turbo-method': 'put' }
+              data: { 'turbo-stream': true, 'turbo-method': 'put' },
+              test_selector: "close-meeting-button"
+
```
```suggestion
```

toy commented on modules/meeting/app/controllers/meeting_agenda_items_controller.rb:
```diff
@@ -199,6 +200,12 @@ def set_meeting
 
+  # In case we updated the meeting as part of the service flow
+  # it needs to be reassigned for the controller in order to get correct timestamps
+  def set_meeting_from_agenda_item
```
Not sure if it helps, but maybe better something like `reset_meeting_from_agenda_item` for clarity that it is set again?

oliverguenther replied:
Changed it 👍

toy commented on modules/meeting/app/models/meeting.rb:
```diff
@@ -110,6 +110,25 @@ class Meeting < ApplicationRecord
+      .connection
+      .exec_query(OpenProject::SqlSanitization.sanitize(sql, id:))
+      .rows
+      .flatten
```
There is no need to fetch `lock_version`, as it is already available on the instance and should be up to date.

And the updated_at values can also be fetched using dsl either in minimum code and two queries:
```rb
[agenda_items.maximum(:updated_at), sections.maximum(:updated_at)]
```
or in one:
```rb
Meeting
  .unscoped # remove default order scope
  .where(id:)
  .left_joins(:agenda_items, :sections)
  .pick("MAX(meeting_agenda_items.updated_at)", "MAX(meeting_sections.updated_at)")
```
or for even more dsl:
```rb
Meeting
  .unscoped
  .where(id:)
  .left_joins(:agenda_items, :sections)
  .pick(MeetingAgendaItem.arel_table[:updated_at].maximum, MeetingSection.arel_table[:updated_at].maximum)
```

toy commented on frontend/src/stimulus/controllers/poll-for-changes.controller.ts:
```diff
@@ -0,0 +1,81 @@
+      }, this.intervalValue || 10_000);
+    }
+
+    this.element.addEventListener('op:poll-for-changes:trigger', this.pollListener);
```
Is the event listener only for tests?

If yes, I think it is better to [directly invoke the controller](https://stimulus.hotwired.dev/reference/controllers#directly-invoking-other-controllers) from test, otherwise listener can be added using `data-action` attribute (probably `op:poll-for-changes:trigger->poll-for-changes#pollListener`)

oliverguenther replied:
Great idea! I read about this once but didn't think about using it for tests

oliverguenther replied:
FYI `this.application` is obviously not available but we expose the stimulus application instance as `window.Stimulus`. Works great

toy commented on frontend/src/stimulus/controllers/poll-for-changes.controller.ts:
```diff
@@ -0,0 +1,81 @@
+    this.element.removeEventListener('op:poll-for-changes:trigger', this.pollListener);
+  }
+
+  async triggerTurboStream():Promise<void> {
```
Why does it need to be async?

oliverguenther replied:
It doesn't, I'll remove it 👍

toy commented on frontend/src/stimulus/controllers/poll-for-changes.controller.ts:
```diff
@@ -0,0 +1,81 @@
+
+  private interval:number;
+
+  private pollListener:() => unknown = () => this.triggerTurboStream();
```
Why not use `triggerTurboStream` directly?

oliverguenther replied:
You need to have the function bound as a variable so that it can be correctly remove it. Just calling `removeEventListener('..', this.myFunction.bind(this)` will not work, as `.bind` returns a new function instance.

toy replied:
I only understood now that `pollListener` captures `this` because it is an arrow function

toy commented on frontend/src/turbo/flash-stream-action.ts:
```diff
@@ -0,0 +1,15 @@
+  StreamActions.flash = function dialogStreamAction(this:StreamElement) {
+    const content = this.templateElement.content;
+
+    const flash = content.querySelector('.flash') as HTMLElement;
+    if (flash.dataset.singleton) {
+      const elements = document.querySelectorAll('.flash');
+      elements.forEach((element) => element.parentElement?.remove());
+    }
```
What does this do?

oliverguenther replied:
We thought about adding some additional logic to replace existing flashes, but that ended up not being necessary. I'll remove it

toy reviewed (changes requested):
Looks good overall, just few notes code wise, also when testing I noticed a problematic issue, as the flash is shown also following my own actions on the page, which will be very confusing, at least it does for adding/removing/renaming sections. I am thinking only of some way of section/agenda item actions to update the checksum in stimulus controller if it is still the same as one that we have.

Was Action cable considered, or is this use "too much" to start using Action cable in the project?

oliverguenther committed:
Replace custom event with accessing controller

oliverguenther committed:
Remove unused flash singleton

oliverguenther committed:
Reload meeting from section similar to agenda

oliverguenther committed:
Switch to click link or button

oliverguenther committed:
Replace inline actions with constant

oliverguenther committed:
Merge branch 'dev' into feature/meetings-update-flash

oliverguenther committed:
Switch to click link or button

oliverguenther committed:
Remove space

oliverguenther committed:
Replace plain SQL with dsl

oliverguenther committed:
Fix reference to INLINE_ACTIONS

oliverguenther commented:
> Looks good overall, just few notes code wise, also when testing I noticed a problematic issue, as the flash is shown also following my own actions on the page, which will be very confusing, at least it does for adding/removing/renaming sections. I am thinking only of some way of section/agenda item actions to update the checksum in stimulus controller if it is still the same as one that we have.

We missed updating the meeting from the service update, which resulted in the incorrect header being updated. This appears to be fixed now, although we are aware with this approach of using dependent values that we'll likely miss more cases where the dialog accidentally triggers. That's one reason why it's still behind a feature flag for now.

> Was Action cable considered, or is this use "too much" to start using Action cable in the project?

We want to introduce ActionCable as soon as we can use the PostgreSQL adapter. I believe that this is now possible with Turbo 8, but someone would have to take a look at the exact details.

oliverguenther committed:
Lint

toy commented:
> > Looks good overall, just few notes code wise, also when testing I noticed a problematic issue, as the flash is shown also following my own actions on the page, which will be very confusing, at least it does for adding/removing/renaming sections. I am thinking only of some way of section/agenda item actions to update the checksum in stimulus controller if it is still the same as one that we have.
> 
> We missed updating the meeting from the service update, which resulted in the incorrect header being updated. This appears to be fixed now, although we are aware with this approach of using dependent values that we'll likely miss more cases where the dialog accidentally triggers. That's one reason why it's still behind a feature flag for now.

I only understood now that the updating I mentioned was already there, done through turbo stream.

MeetingSectionsController has `update_header_component_via_turbo_stream` only in `update` and `move`, so `create`, `destroy` and `drop` still need it.

Also I noticed that `MeetingAgendaItemsController#move` calls `move_item_within_section_via_turbo_stream` and `update_header_component_via_turbo_stream`, but `move_item_within_section_via_turbo_stream` already includes call to `update_header_component_via_turbo_stream`.

toy commented:
There is still a way to miss changes due to polling which can be easily displayed by moving sections, and I also saw a problem with the state of sections.

1. Create 3 sections (named `a`, `b`, `c`) and open it in two windows
2. Next should be done quickly, in first window tell section `a` to move to bottom, and in the second window tell section `b` to move to bottom
3. After poll first window gets the reload flash, but second does not

Problem with state is that in second window despite the order being `a, c, b`, the actions for moving are correctly updated, so section `a` shows actions to move to top and bottom, section `c` shows only actions to move to bottom and section `b` shows only actions to move to top.

mrmir committed:
Edit update_header_component calls

oliverguenther commented:
@toy we address the remarks regarding the header. For the quick subsequent reorders, we opened an open point since it's inherent to the way section reordering works with acts_as_list (not sending the full order to the backend, but only updating parts of it).

Discussion for that continues in https://community.openproject.org/work_packages/57331

If you agree, we'd like to merge this in

mrmir committed:
Merge branch 'dev' into feature/meetings-update-flash

oliverguenther committed:
Merge branch 'dev' into feature/meetings-update-flash

mrmir committed:
Lint setup.ts

toy commented on modules/meeting/app/components/meetings/header_component.html.erb:
```diff
@@ -1,6 +1,15 @@
+    render(Primer::OpenProject::PageHeader.new(
+      test_selector: "meeting-page-header",
+      state: @state,
+      data: {
+        controller: "poll-for-changes",
+        poll_for_changes_reference_value: @meeting.changed_hash,
+        poll_for_changes_url_value: check_for_updates_meeting_path(@meeting),
+        poll_for_changes_interval_value: check_for_updates_interval,
+      }
```
Not adding this data unless `OpenProject::FeatureDecisions.meeting_updated_notification_active?` will really include less of the feature in running code

toy commented:
I agree, it is anyway behind a feature flag. I commented on the way it is added when checking on it

mrmir committed:
Update header_component to conditionally send data

mrmir committed:
Remove redundant merge
